core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java

```diff
@@ -0,0 +1,274 @@
+    currentOffsetInPlainBlock = Math.toIntExact(newPos % plainBlockSize);
+    plainStreamPosition = newPos;
+  }
+
+  @Override
+  public long skip(long n) {
+    if (n <= 0) {
+      return 0;
+    }
+
+    if (plainStreamPosition == plainStreamSize) {
+      return 0;
+    }
+
+    long newPosition = plainStreamPosition + n;
+
+    if (newPosition > plainStreamSize) {
+      long skipped = plainStreamSize - plainStreamPosition;
+      try {
+        seek(plainStreamSize);
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+      return skipped;
+    }
+
+    try {
+      seek(newPosition);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+    return n;
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    return plainStreamPosition;
+  }
+
+  @Override
+  public int read() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void close() throws IOException {
+    sourceStream.close();
+    currentDecryptedBlock = null;
+  }
+
+  static long calculatePlaintextLength(long sourceLength, int plainBlockSize) {
+    long netSourceFileLength = netSourceFileLength(sourceLength);
+
+    if (netSourceFileLength == 0) {
+      return 0;
+    }
+
+    int cipherBlockSize = plainBlockSize + Ciphers.NONCE_LENGTH + Ciphers.GCM_TAG_LENGTH;
+    int numberOfFullBlocks = Math.toIntExact(netSourceFileLength / cipherBlockSize);
+    int cipherBytesInLastBlock =
+        Math.toIntExact(netSourceFileLength - numberOfFullBlocks * cipherBlockSize);
+    boolean fullBlocksOnly = (0 == cipherBytesInLastBlock);
+    int plainBytesInLastBlock =
+        fullBlocksOnly
+            ? 0
+            : (cipherBytesInLastBlock - Ciphers.NONCE_LENGTH - Ciphers.GCM_TAG_LENGTH);
+
+    return (long) numberOfFullBlocks * plainBlockSize + plainBytesInLastBlock;
+  }
+
+  private byte[] decryptNextBlock(boolean isLastBlockInStream) throws IOException {
+    if (currentBlockIndex == currentDecryptedBlockIndex) {
+      return currentDecryptedBlock;
+    }
+
+    long blockPositionInStream = blockOffset(currentBlockIndex);
+    if (sourceStream.getPos() != blockPositionInStream) {
+      sourceStream.seek(blockPositionInStream);
+    }
+
+    int currentCipherBlockSize = isLastBlockInStream ? lastCipherBlockSize : cipherBlockSize;
+    IOUtil.readFully(sourceStream, cipherBlockBuffer, 0, currentCipherBlockSize);
+
+    byte[] aad = Ciphers.streamBlockAAD(fileAADPrefix, currentBlockIndex);
+    byte[] result = gcmDecryptor.decrypt(cipherBlockBuffer, 0, currentCipherBlockSize, aad);
+    currentDecryptedBlockIndex = currentBlockIndex;
+    currentDecryptedBlock = result;
+    return result;
+  }
+
+  private long blockOffset(int blockIndex) {
+    return (long) blockIndex * cipherBlockSize + Ciphers.GCM_STREAM_HEADER_LENGTH;
+  }
+
+  private static long netSourceFileLength(long sourceFileLength) {
```

Comment:
I don't think that "net" is clear enough, and I also think that this method is doing too much. A calculation should not include a Precondition check like this because it is surprising. You must know the implementation details of the method to call it properly.

I think that the length should be validated in `newStream` when attempting to read the file, and _not_ when calculating the file length for `InputStream.getLength`.

I think checking at that point avoids the need to have this method entirely, which avoids the naming problem. The call can simply be `encryptedFileLength - Ciphers.GCM_STREAM_HEADER_LENGTH`.